Threaded comments

tavinathanson opened the pull request:
Most of the changes:
- Only show the VCFCommentHeader at the top of a thread.
- Default to edit mode ("Enter your comment here") for the comment at the bottom of the thread.
- Show author name and humanized created dates for each comment, and sort comments by that date.
- Use localStorage to store author name until a new author name is entered, and allow anonymous commenting. 
- Add comment count to each row.
- Add author name to news feed.
- React-ified the /comments page.
- Restricted some typeahead CSS to be more specific (artifact of using typeahead and ripping it out, but seems like a reasonable change).
- Updated `ExaminePage-comments-test.js`, including a small multiple-comments test.

TODO:
- Use `seltest`?
- More tests?
- GitHub-like styling (another PR).

Migration:

```
alter table user_comments add column author_name TEXT;
alter table user_comments add column created TIMESTAMP NOT NULL DEFAULT CURRENT_TIMESTAMP;
```

![screen shot 2015-03-23 at 7 31 32 pm](https://cloud.githubusercontent.com/assets/322932/6792854/46c5e0b8-d193-11e4-846d-bff9db58972b.png)

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/hammerlab/cycledash/553)

<!-- Reviewable:end -->

tavinathanson committed:
A few author-name related changes

tavinathanson committed:
Better styling, humanized dates, and other fixes

tavinathanson committed:
Update comments testing DB

tavinathanson committed:
Changes to the feed and comment counts

tavinathanson committed:
Fix comments testing utils and tests

tavinathanson committed:
Go back to updating commentMap

danvk commented:
Cool!

@ihodes has a fix for Travis that should go in as part of #548

danvk commented:
@tavinathanson mind rebasing off of `master` so you can get the Travis fix?

tavinathanson committed:
A first pass at multiple comments per record

tavinathanson committed:
Query for comments every time, and bug fix

tavinathanson committed:
Remove a sinon check and an invalid function call

tavinathanson committed:
Fix linter error

tavinathanson commented:
@danvk rebased and squashed

danvk commented:
Cool, taking a look shortly...

tavinathanson commented:
:+1: no rush

tavinathanson committed:
Updating data.sql and screenshots

tavinathanson commented:
(Not sure how `TEST_PROJECT` got into `data.sql`, I'll remove it...)

danvk commented:
Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/hammerlab/cycledash/553)

Reviewed files:
- cycledash/comments.py @ r5
- cycledash/static/css/examine.css @ r5
- cycledash/static/css/runs.css @ r5
- cycledash/static/css/style.css @ r5
- schema.sql @ r5
- tests/data/comment-db.json @ r5
- tests/js/CommentUtils.js @ r5
- tests/js/ExaminePage-comments-test.js @ r6
- tests/pdifftests/test.py @ r7

---

Not a super thorough review, but here are a few comments to get you started. Also, it looks like Travis is broken again?

---

<sup>**[cycledash/static/js/examine/components/CommentBox.js, line 194 \[r7\]](https://reviewable.io:443/reviews/hammerlab/cycledash/553#-JlDCXxAGspn1Fp98yhy)** ([raw file](https://github.com/hammerlab/cycledash/blob/f6e0a142f7058a904084d2028af03c9ac68d5797/cycledash/static/js/examine/components/CommentBox.js#L194)):</sup>
reads better as `isEditing`?

---

<sup>**[cycledash/static/js/examine/components/CommentBox.js, line 225 \[r7\]](https://reviewable.io:443/reviews/hammerlab/cycledash/553#-JlDCQ3-XOSOCK6XCQgR)** ([raw file](https://github.com/hammerlab/cycledash/blob/f6e0a142f7058a904084d2028af03c9ac68d5797/cycledash/static/js/examine/components/CommentBox.js#L225)):</sup>
You can just use `this.props.authorName || 'Anonymous'`

---

<sup>**[cycledash/static/js/examine/RecordStore.js, line 214 \[r7\]](https://reviewable.io:443/reviews/hammerlab/cycledash/553#-JlDDphVhXAUwwdDUpW1)** ([raw file](https://github.com/hammerlab/cycledash/blob/f6e0a142f7058a904084d2028af03c9ac68d5797/cycledash/static/js/examine/RecordStore.js#L214)):</sup>
we have to do this here as well as in SQL for optimistic updating?

---

<sup>**[tests/js/CommentUtils.js, line 79 \[r7\]](https://reviewable.io:443/reviews/hammerlab/cycledash/553#-JlDB3n2uJCBFyg30p9d)** ([raw file](https://github.com/hammerlab/cycledash/blob/f6e0a142f7058a904084d2028af03c9ac68d5797/tests/js/CommentUtils.js#L79)):</sup>
Looks like you've implemented `_.groupBy` :)

```javascript
var obj = _.groupBy(commentDatabase, utils.getRowKey)
```

---

<sup>**[tests/js/ExaminePage-comments-test.js, line 62 \[r7\]](https://reviewable.io:443/reviews/hammerlab/cycledash/553#-JlDBJOZ6vUu3AUCRKeH)** ([raw file](https://github.com/hammerlab/cycledash/blob/f6e0a142f7058a904084d2028af03c9ac68d5797/tests/js/ExaminePage-comments-test.js#L62)):</sup>
document what `n` is (and below), e.g. `// get text for the nth comment`

---

<sup>**[tests/js/ExaminePage-comments-test.js, line 64 \[r7\]](https://reviewable.io:443/reviews/hammerlab/cycledash/553#-JlDBXLAXnHQ8KCgu4IP)** ([raw file](https://github.com/hammerlab/cycledash/blob/f6e0a142f7058a904084d2028af03c9ac68d5797/tests/js/ExaminePage-comments-test.js#L64)):</sup>
Why 1-based?

---

<sup>**[tests/pdifftests/test.py, line 46 \[r7\]](https://reviewable.io:443/reviews/hammerlab/cycledash/553#-JlDBc5q_Mau7JHK31VG)** ([raw file](https://github.com/hammerlab/cycledash/blob/f6e0a142f7058a904084d2028af03c9ac68d5797/tests/pdifftests/test.py#L46)):</sup>
was this something you guys needed to change to get it to work on your machine? (I'd like to avoid changing dimensions of the screenshots if at all possible, since it makes them un-diffable)

---


<!-- Sent from Reviewable.io -->

tavinathanson committed:
Address code review

tavinathanson commented:
Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/hammerlab/cycledash/553)







---

<sup>**[cycledash/static/js/examine/components/CommentBox.js, line 194 \[r7\]](https://reviewable.io:443/reviews/hammerlab/cycledash/553#-JlDCXxAGspn1Fp98yhy)** ([raw file](https://github.com/hammerlab/cycledash/blob/f6e0a142f7058a904084d2028af03c9ac68d5797/cycledash/static/js/examine/components/CommentBox.js#L194)):</sup>
Done

---

<sup>**[cycledash/static/js/examine/components/CommentBox.js, line 225 \[r7\]](https://reviewable.io:443/reviews/hammerlab/cycledash/553#-JlDCQ3-XOSOCK6XCQgR)** ([raw file](https://github.com/hammerlab/cycledash/blob/f6e0a142f7058a904084d2028af03c9ac68d5797/cycledash/static/js/examine/components/CommentBox.js#L225)):</sup>
Done

---

<sup>**[cycledash/static/js/examine/RecordStore.js, line 214 \[r7\]](https://reviewable.io:443/reviews/hammerlab/cycledash/553#-JlDDphVhXAUwwdDUpW1)** ([raw file](https://github.com/hammerlab/cycledash/blob/f6e0a142f7058a904084d2028af03c9ac68d5797/cycledash/static/js/examine/RecordStore.js#L214)):</sup>
I don't think I understand the question? Do you mean `updateCommentInParentRecord` in general? Without this, the comments aren't present inside the record objects.

Or are you asking about deletes? Or something else?

---

<sup>**[tests/js/CommentUtils.js, line 79 \[r7\]](https://reviewable.io:443/reviews/hammerlab/cycledash/553#-JlDB3n2uJCBFyg30p9d)** ([raw file](https://github.com/hammerlab/cycledash/blob/f6e0a142f7058a904084d2028af03c9ac68d5797/tests/js/CommentUtils.js#L79)):</sup>
:)

---

<sup>**[tests/js/ExaminePage-comments-test.js, line 62 \[r7\]](https://reviewable.io:443/reviews/hammerlab/cycledash/553#-JlDBJOZ6vUu3AUCRKeH)** ([raw file](https://github.com/hammerlab/cycledash/blob/f6e0a142f7058a904084d2028af03c9ac68d5797/tests/js/ExaminePage-comments-test.js#L62)):</sup>
Good call. I was just being lazy!

---

<sup>**[tests/js/ExaminePage-comments-test.js, line 64 \[r7\]](https://reviewable.io:443/reviews/hammerlab/cycledash/553#-JlDBXLAXnHQ8KCgu4IP)** ([raw file](https://github.com/hammerlab/cycledash/blob/f6e0a142f7058a904084d2028af03c9ac68d5797/tests/js/ExaminePage-comments-test.js#L64)):</sup>
No particular reason. Made it 0-based now.

---

<sup>**[tests/pdifftests/test.py, line 46 \[r7\]](https://reviewable.io:443/reviews/hammerlab/cycledash/553#-JlDBc5q_Mau7JHK31VG)** ([raw file](https://github.com/hammerlab/cycledash/blob/f6e0a142f7058a904084d2028af03c9ac68d5797/tests/pdifftests/test.py#L46)):</sup>
Using phantom 2, my screenshots were showing up with different dimensions than what was specified here. If I set the dimensions to lower than 1796w, I just get screenshots that are 1796w. Higher than 1796w and I get screenshots that meet the specification. This was just the quickest thing to do here, but open to suggestions.

---


<!-- Sent from Reviewable.io -->

danvk commented:
Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/hammerlab/cycledash/553)



---

Maybe @ihodes could take a look at this? I need to hunker down on my Python Meetup preso today (I'm giving it tomorrow).

---

<sup>**[tests/pdifftests/test.py, line 46 \[r7\]](https://reviewable.io:443/reviews/hammerlab/cycledash/553#-JlDBc5q_Mau7JHK31VG)** ([raw file](https://github.com/hammerlab/cycledash/blob/f6e0a142f7058a904084d2028af03c9ac68d5797/tests/pdifftests/test.py#L46)):</sup>
We should figure out how to fix that—changing the dimensions of a screenshot is a pretty drastic thing to do, since it makes it impossible to look at the before/after diff. 1800px seems like a crazy large minimum. What if you want to test a mobile version of the site?

---


<!-- Sent from Reviewable.io -->

ihodes commented:
Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/hammerlab/cycledash/553)







---

<sup>**[tests/pdifftests/test.py, line 46 \[r7\]](https://reviewable.io:443/reviews/hammerlab/cycledash/553#-JlDBc5q_Mau7JHK31VG)** ([raw file](https://github.com/hammerlab/cycledash/blob/f6e0a142f7058a904084d2028af03c9ac68d5797/tests/pdifftests/test.py#L46)):</sup>
I believe that this is a function of how the browser takes screenshots; phantomjs version 2.0 seems to (I haven't run a before-after test on my machine; this is the behavior we're seeing with v2 that we didn't see with v1, though) restrict the screenshot to the viewport, similar to how other browsers (e.g. Firefox iirc) do. 

If we wanted to test a mobile version of the website, we could have another test suite that sets the browser size to some other dimension, and then take smaller screenshots; that's not a problem at all (and even if it were, something I don't think we need to worry about with CycleDash).

I don't think that we need to worry about the changing size of the screenshots; viewing the diffs is still easy (just blink or view the images size-by-side), and the sizes change rarely. It's not arduous at all to review the screenshots every time we're forced to change the dimensions; we're not worried about small pixel changes anyway, just that the desired functionality is being displayed in the screenshot.

---


<!-- Sent from Reviewable.io -->

danvk commented:
Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/hammerlab/cycledash/553)







---

<sup>**[tests/pdifftests/test.py, line 46 \[r7\]](https://reviewable.io:443/reviews/hammerlab/cycledash/553#-JlDBc5q_Mau7JHK31VG)** ([raw file](https://github.com/hammerlab/cycledash/blob/f6e0a142f7058a904084d2028af03c9ac68d5797/tests/pdifftests/test.py#L46)):</sup>
Interesting, I'm not sure if I see this when I use PhantomJS2 with dpxdt. Maybe you guys can help me run seltest on my machine with Phantom2.

I disagree that we don't care about small, unexpected changes, but let's table that discussion for this review.

If we need to change the size of the screenshots, we should do it in a separate PR that goes in before this one and changes the sizes and nothing else (like I did with the rename). That way you'll get clearer diffs for real code changes (like this PR).

---


<!-- Sent from Reviewable.io -->

ihodes commented:
Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/hammerlab/cycledash/553)







---

<sup>**[tests/pdifftests/test.py, line 46 \[r7\]](https://reviewable.io:443/reviews/hammerlab/cycledash/553#-JlDBc5q_Mau7JHK31VG)** ([raw file](https://github.com/hammerlab/cycledash/blob/f6e0a142f7058a904084d2028af03c9ac68d5797/tests/pdifftests/test.py#L46)):</sup>
> I disagree that we don't care about small, unexpected changes, but let's table that discussion for this review.

That's definitely not what I intended—I also would want to catch unexpected changes.

> Interesting, I'm not sure if I see this when I use PhantomJS2 with dpxdt. Maybe you guys can help me run seltest on my machine with Phantom2.

I have't tested this theory on machine, as noted, just was a guess. The screenshot capabilities of the various phantomjs versions should be the same between seltest and dpxdt, as they're using the browser's native capabilities. 

> That way you'll get clearer diffs for real code changes (like this PR).

This seems a bit excessive. A visual review indicates any changes and will catch regressions; this seems unnecessary (especially given that a visual review of the screenshots will be required even if you do the pre-merge screenshot-resolution-change-PR).

---


<!-- Sent from Reviewable.io -->

ihodes commented:
Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/hammerlab/cycledash/553)

Reviewed files:
- cycledash/static/css/examine.css @ r5
- cycledash/static/css/runs.css @ r5
- cycledash/static/css/style.css @ r5
- cycledash/static/js/examine/components/CommentBox.js @ r8
- cycledash/static/js/examine/components/ExaminePage.js @ r5
- cycledash/static/js/examine/components/VCFTable.js @ r5
- cycledash/static/js/examine/RecordActions.js @ r5
- cycledash/static/js/examine/RecordStore.js @ r6
- cycledash/static/js/examine/utils.js @ r5
- cycledash/static/js/runs/comments.js @ r5
- cycledash/static/js/runs/components/comment.js @ r5
- cycledash/static/js/runs/components/LatestComments.js @ r5
- cycledash/static/js/runs/components/RunsPage.js @ r5
- cycledash/templates/comments.html @ r5
- cycledash/templates/macros/comments.html @ r5
- cycledash/templates/macros/nav.html @ r5
- gulpfile.js @ r5
- package.json @ r5
- schema.sql @ r5
- tests/data.sql @ r8
- tests/data/comment-db.json @ r5
- tests/js/CommentUtils.js @ r8
- tests/js/ExaminePage-comments-test.js @ r8
- tests/pdifftests/images/examine_bad-query.png @ r7
- tests/pdifftests/images/examine_base.png @ r7
- tests/pdifftests/images/examine_comments-edit.png @ r8
- tests/pdifftests/images/examine_comments-view.png @ r8
- tests/pdifftests/images/examine_filter.png @ r7
- tests/pdifftests/images/examine_sorted.png @ r7
- tests/pdifftests/images/examine_tooltip.png @ r7
- tests/pdifftests/images/examine_validation.png @ r7
- tests/pdifftests/images/runs_bams.png @ r8
- tests/pdifftests/images/runs_info.png @ r8
- tests/pdifftests/images/runs_page.png @ r8
- tests/pdifftests/images/website_about-page.png @ r7
- tests/pdifftests/images/website_comments.png @ r8
- tests/pdifftests/test.py @ r7

---

Back to you, Tavi! This looks good.

---

<sup>**[cycledash/comments.py, line 262 \[r8\]](https://reviewable.io:443/reviews/hammerlab/cycledash/553#-JlHkQ0RCyZuo_kBbw9p)** ([raw file](https://github.com/hammerlab/cycledash/blob/2d190e8e0a9dac68f0a3209ece6e879ac1b023fb/cycledash/comments.py#L262)):</sup>
this method is equivalent to `d.get(key)` :)

---

<sup>**[cycledash/comments.py, line 276 \[r8\]](https://reviewable.io:443/reviews/hammerlab/cycledash/553#-JlHk9m2snyQRSeaLRg8)** ([raw file](https://github.com/hammerlab/cycledash/blob/2d190e8e0a9dac68f0a3209ece6e879ac1b023fb/cycledash/comments.py#L276)):</sup>
instead of the `\`-continuation, how about

```python
convert_col_name(name, user_comments): convert_col_value(
    name, col_name_to_val[name], user_comments)
```

---

<sup>**[cycledash/static/js/examine/components/CommentBox.js, line 90 \[r8\]](https://reviewable.io:443/reviews/hammerlab/cycledash/553#-JlIa6LAOB4Y9OWzHvay)** ([raw file](https://github.com/hammerlab/cycledash/blob/2d190e8e0a9dac68f0a3209ece6e879ac1b023fb/cycledash/static/js/examine/components/CommentBox.js#L90)):</sup>
could just be `LOCAL_STORAGE_AUTHOR_KEY = 'CYCLEDASH_AUTHORNAME';`

---

<sup>**[cycledash/static/js/examine/components/CommentBox.js, line 94 \[r8\]](https://reviewable.io:443/reviews/hammerlab/cycledash/553#-JlI_vTO52KrKHBjQsd3)** ([raw file](https://github.com/hammerlab/cycledash/blob/2d190e8e0a9dac68f0a3209ece6e879ac1b023fb/cycledash/static/js/examine/components/CommentBox.js#L94)):</sup>
Do modern browsers not support local storage? Not that this is a bad thing to do anyway, but if it's just the case for older browsers I don't think we'd care. 

Then, if you don't worry about the checking, this whole function could just be 
```js
function () { 
    return localStorage.getItem(this.LOCAL_STORAGE_AUTHOR_KEY); || '';
}
```

---

<sup>**[cycledash/static/js/examine/components/CommentBox.js, line 103 \[r8\]](https://reviewable.io:443/reviews/hammerlab/cycledash/553#-JlIdqKXZpzH8Aodxhx1)** ([raw file](https://github.com/hammerlab/cycledash/blob/2d190e8e0a9dac68f0a3209ece6e879ac1b023fb/cycledash/static/js/examine/components/CommentBox.js#L103)):</sup>
This could go into VCFCommentEditor, and that way you don't need to pass it down 3 layers of props?

---

<sup>**[cycledash/static/js/examine/components/CommentBox.js, line 114 \[r8\]](https://reviewable.io:443/reviews/hammerlab/cycledash/553#-JlIbI7WHIbyVU4sIEh2)** ([raw file](https://github.com/hammerlab/cycledash/blob/2d190e8e0a9dac68f0a3209ece6e879ac1b023fb/cycledash/static/js/examine/components/CommentBox.js#L114)):</sup>
(Assuming you want to keep this function, which is quite likely; you may very well need/want it) Is there a particular error you're worried about throwing here?

---

<sup>**[cycledash/static/js/examine/components/CommentBox.js, line 129 \[r8\]](https://reviewable.io:443/reviews/hammerlab/cycledash/553#-JlIbfC8y3hCOPkkohsG)** ([raw file](https://github.com/hammerlab/cycledash/blob/2d190e8e0a9dac68f0a3209ece6e879ac1b023fb/cycledash/static/js/examine/components/CommentBox.js#L129)):</sup>
You knew this would happen! :) I think this might be a bit more JS-y with a `.map` like so, let me know what you think. This save you the pre-loop initialization, the use of underscore for `_.each` (`Array.map` is built in!) And you don't have to mutate state within a loop (my personal favorite!) 

```js
var commentNodes = _.sortBy(comments, c => { new Date(c.created_date).getTime() }
).map(c => {
  // etc…
  return <VCFComment …>;
});
```

Also, camelCase your vars (the perils of going between .py and .js)

---

<sup>**[cycledash/static/js/examine/components/CommentBox.js, line 183 \[r8\]](https://reviewable.io:443/reviews/hammerlab/cycledash/553#-JlIdUzvXmyQcW6eUfUR)** ([raw file](https://github.com/hammerlab/cycledash/blob/2d190e8e0a9dac68f0a3209ece6e879ac1b023fb/cycledash/static/js/examine/components/CommentBox.js#L183)):</sup>
`editable` seems more boolean-y to me?

---

<sup>**[cycledash/static/js/examine/components/CommentBox.js, line 184 \[r8\]](https://reviewable.io:443/reviews/hammerlab/cycledash/553#-JlIdYgEFm41bjGWOJ72)** ([raw file](https://github.com/hammerlab/cycledash/blob/2d190e8e0a9dac68f0a3209ece6e879ac1b023fb/cycledash/static/js/examine/components/CommentBox.js#L184)):</sup>
similarly w/ `cancelable` 

---

<sup>**[cycledash/static/js/examine/components/CommentBox.js, line 209 \[r8\]](https://reviewable.io:443/reviews/hammerlab/cycledash/553#-JlIe54oB9XXq6v2bsb7)** ([raw file](https://github.com/hammerlab/cycledash/blob/2d190e8e0a9dac68f0a3209ece6e879ac1b023fb/cycledash/static/js/examine/components/CommentBox.js#L209)):</sup>
`this.setState({isEditing})`

---

<sup>**[cycledash/static/js/examine/components/CommentBox.js, line 330 \[r8\]](https://reviewable.io:443/reviews/hammerlab/cycledash/553#-JlIee9xTBVJG6V0jvM-)** ([raw file](https://github.com/hammerlab/cycledash/blob/2d190e8e0a9dac68f0a3209ece6e879ac1b023fb/cycledash/static/js/examine/components/CommentBox.js#L330)):</sup>
livin' on the edge :) 

---

<sup>**[cycledash/static/js/examine/components/CommentBox.js, line 373 \[r8\]](https://reviewable.io:443/reviews/hammerlab/cycledash/553#-JlLUPyp22PLbJA12tj2)** ([raw file](https://github.com/hammerlab/cycledash/blob/2d190e8e0a9dac68f0a3209ece6e879ac1b023fb/cycledash/static/js/examine/components/CommentBox.js#L373)):</sup>
no need for `return;`

---

<sup>**[cycledash/static/js/examine/components/VCFTable.js, line 258 \[r8\]](https://reviewable.io:443/reviews/hammerlab/cycledash/553#-JlIetrTxEryVMYGHSEV)** ([raw file](https://github.com/hammerlab/cycledash/blob/2d190e8e0a9dac68f0a3209ece6e879ac1b023fb/cycledash/static/js/examine/components/VCFTable.js#L258)):</sup>
Good call

---

<sup>**[cycledash/static/js/examine/RecordStore.js, line 214 \[r7\]](https://reviewable.io:443/reviews/hammerlab/cycledash/553#-JlDDphVhXAUwwdDUpW1)** ([raw file](https://github.com/hammerlab/cycledash/blob/f6e0a142f7058a904084d2028af03c9ac68d5797/cycledash/static/js/examine/RecordStore.js#L214)):</sup>
I have a similar/the same question, I think; if we could do the join on the server-side (in SQL), we could get rid of a lot of careful client-side code that's doing explicit joins between comments & records.

---

<sup>**[cycledash/static/js/examine/RecordStore.js, line 215 \[r8\]](https://reviewable.io:443/reviews/hammerlab/cycledash/553#-JlLYRjXLbvEHzTGr2jt)** ([raw file](https://github.com/hammerlab/cycledash/blob/2d190e8e0a9dac68f0a3209ece6e879ac1b023fb/cycledash/static/js/examine/RecordStore.js#L215)):</sup>
This branch could be done with two lines of underscore using `_.without`, e.g. 

```js
record.comments = _.without(records.comments, comment);
return comment;
```

---

<sup>**[cycledash/static/js/examine/RecordStore.js, line 229 \[r8\]](https://reviewable.io:443/reviews/hammerlab/cycledash/553#-JlL_gr5yJtZjyg8daM7)** ([raw file](https://github.com/hammerlab/cycledash/blob/2d190e8e0a9dac68f0a3209ece6e879ac1b023fb/cycledash/static/js/examine/RecordStore.js#L229)):</sup>
Could replace the `each` here and save a bit of code, let me know what  you think (this seems a bit clearer to me).

```js
if (_.has(comment, 'id')) {
    var oldComment = _.findWhere(record.comments, {id: comment.id});
} else {
    var oldComment = _.findWhere(record.comments, {created_date: comment.created_date});
}

if (oldComment) {
    var clonedComment = _.clone(oldComment);  // NB _.clone is shallow.
    _.extend(oldComment, comment);  // replace values in oldComment 
    return clonedComment;
} else {
    record.comments = record.comments || [];
    record.comments.push(comment);
}
```

---

<sup>**[cycledash/static/js/examine/RecordStore.js, line 239 \[r8\]](https://reviewable.io:443/reviews/hammerlab/cycledash/553#-JlLWLXEDWgUY8VbobSr)** ([raw file](https://github.com/hammerlab/cycledash/blob/2d190e8e0a9dac68f0a3209ece6e879ac1b023fb/cycledash/static/js/examine/RecordStore.js#L239)):</sup>
"is its"?

---

<sup>**[cycledash/static/js/examine/RecordStore.js, line 334 \[r8\]](https://reviewable.io:443/reviews/hammerlab/cycledash/553#-JlLcaorg1BOmgECTfwE)** ([raw file](https://github.com/hammerlab/cycledash/blob/2d190e8e0a9dac68f0a3209ece6e879ac1b023fb/cycledash/static/js/examine/RecordStore.js#L334)):</sup>
Not necessary to change, but there's an underscore function for this! 

```js
_.pick(comment, 'author_name', 'last_modified_timestamp', 'comment_text');
```

---

<sup>**[cycledash/static/js/examine/utils.js, line 62 \[r8\]](https://reviewable.io:443/reviews/hammerlab/cycledash/553#-JlLd5cYD2a1QpKsQ6KM)** ([raw file](https://github.com/hammerlab/cycledash/blob/2d190e8e0a9dac68f0a3209ece6e879ac1b023fb/cycledash/static/js/examine/utils.js#L62)):</sup>
indent following lines by 2 (I think it's 2!)

---

<sup>**[cycledash/static/js/runs/comments.js, line 1 \[r8\]](https://reviewable.io:443/reviews/hammerlab/cycledash/553#-JlLdLuDGrU_NVJjDPmE)** ([raw file](https://github.com/hammerlab/cycledash/blob/2d190e8e0a9dac68f0a3209ece6e879ac1b023fb/cycledash/static/js/runs/comments.js#L1)):</sup>
This (and other comments-page related view code) should probably go into their own static/js/comments directory now, since comments have their own page now. 

---

<sup>**[package.json, line 9 \[r8\]](https://reviewable.io:443/reviews/hammerlab/cycledash/553#-JlLe-9Styl6XjoisS2H)** ([raw file](https://github.com/hammerlab/cycledash/blob/2d190e8e0a9dac68f0a3209ece6e879ac1b023fb/package.json#L9)):</sup>
We're using moment.js for this now—cf. `LatestComments` (grep for `relativeDate`), so I don't think we need humanize.

---

<sup>**[schema.sql, line 47 \[r8\]](https://reviewable.io:443/reviews/hammerlab/cycledash/553#-JlLeBu86a97bz0hZREA)** ([raw file](https://github.com/hammerlab/cycledash/blob/2d190e8e0a9dac68f0a3209ece6e879ac1b023fb/schema.sql#L47)):</sup>
Could you make a note of migration code in the PR on GitHub?

---

<sup>**[tests/data/comment-db.json, line 46 \[r8\]](https://reviewable.io:443/reviews/hammerlab/cycledash/553#-JlLflxkHizTGMY__G6m)** ([raw file](https://github.com/hammerlab/cycledash/blob/2d190e8e0a9dac68f0a3209ece6e879ac1b023fb/tests/data/comment-db.json#L46)):</sup>
Hmm is this how we're storing datetimes? Using ISO 8601-style datetime+TZ seems more robust & easily parseable. Not a big deal if the change is large—this is regular enough.

---


<!-- Sent from Reviewable.io -->

tavinathanson committed:
Address more code review comments

tavinathanson commented:
Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/hammerlab/cycledash/553)



---

Also noticed that /comments had the same timezone issues that I addressed in CommentBox, so I added the timezone offset to /comments as well.

I'm tracking down a test failure and didn't yet resolve the screenshot debate.

Other than those items, back to you!

---

<sup>**[cycledash/comments.py, line 262 \[r8\]](https://reviewable.io:443/reviews/hammerlab/cycledash/553#-JlHkQ0RCyZuo_kBbw9p)** ([raw file](https://github.com/hammerlab/cycledash/blob/2d190e8e0a9dac68f0a3209ece6e879ac1b023fb/cycledash/comments.py#L262)):</sup>
Fixed :)

---

<sup>**[cycledash/static/js/examine/components/CommentBox.js, line 90 \[r8\]](https://reviewable.io:443/reviews/hammerlab/cycledash/553#-JlIa6LAOB4Y9OWzHvay)** ([raw file](https://github.com/hammerlab/cycledash/blob/2d190e8e0a9dac68f0a3209ece6e879ac1b023fb/cycledash/static/js/examine/components/CommentBox.js#L90)):</sup>
Done

---

<sup>**[cycledash/static/js/examine/components/CommentBox.js, line 94 \[r8\]](https://reviewable.io:443/reviews/hammerlab/cycledash/553#-JlI_vTO52KrKHBjQsd3)** ([raw file](https://github.com/hammerlab/cycledash/blob/2d190e8e0a9dac68f0a3209ece6e879ac1b023fb/cycledash/static/js/examine/components/CommentBox.js#L94)):</sup>
Looks like we also (or, at least) need to avoid private browsing mode issues! Gross. http://stackoverflow.com/questions/14555347/html5-localstorage-error-with-safari-quota-exceeded-err-dom-exception-22-an

I'm going to use store.js, which seems to support that use case cleanly. https://github.com/marcuswestin/store.js

---

<sup>**[cycledash/static/js/examine/components/CommentBox.js, line 103 \[r8\]](https://reviewable.io:443/reviews/hammerlab/cycledash/553#-JlIdqKXZpzH8Aodxhx1)** ([raw file](https://github.com/hammerlab/cycledash/blob/2d190e8e0a9dac68f0a3209ece6e879ac1b023fb/cycledash/static/js/examine/components/CommentBox.js#L103)):</sup>
Hmm. I hear you, but it feels less confusing to me to have all set/get authorName logic in the same place.

---

<sup>**[cycledash/static/js/examine/components/CommentBox.js, line 114 \[r8\]](https://reviewable.io:443/reviews/hammerlab/cycledash/553#-JlIbI7WHIbyVU4sIEh2)** ([raw file](https://github.com/hammerlab/cycledash/blob/2d190e8e0a9dac68f0a3209ece6e879ac1b023fb/cycledash/static/js/examine/components/CommentBox.js#L114)):</sup>
Now using store.js. (But an error on private browsing comes to mind.)

---

<sup>**[cycledash/static/js/examine/components/CommentBox.js, line 129 \[r8\]](https://reviewable.io:443/reviews/hammerlab/cycledash/553#-JlIbfC8y3hCOPkkohsG)** ([raw file](https://github.com/hammerlab/cycledash/blob/2d190e8e0a9dac68f0a3209ece6e879ac1b023fb/cycledash/static/js/examine/components/CommentBox.js#L129)):</sup>
Very well, then! Except I like `comment` over `c`.

---

<sup>**[cycledash/static/js/examine/components/CommentBox.js, line 183 \[r8\]](https://reviewable.io:443/reviews/hammerlab/cycledash/553#-JlIdUzvXmyQcW6eUfUR)** ([raw file](https://github.com/hammerlab/cycledash/blob/2d190e8e0a9dac68f0a3209ece6e879ac1b023fb/cycledash/static/js/examine/components/CommentBox.js#L183)):</sup>
Except this doesn't mean that. It means that it specifically starts out in the edit state. I'll change to `startInEditState` to be more clear?

---

<sup>**[cycledash/static/js/examine/components/CommentBox.js, line 184 \[r8\]](https://reviewable.io:443/reviews/hammerlab/cycledash/553#-JlIdYgEFm41bjGWOJ72)** ([raw file](https://github.com/hammerlab/cycledash/blob/2d190e8e0a9dac68f0a3209ece6e879ac1b023fb/cycledash/static/js/examine/components/CommentBox.js#L184)):</sup>
Done.

---

<sup>**[cycledash/static/js/examine/components/CommentBox.js, line 209 \[r8\]](https://reviewable.io:443/reviews/hammerlab/cycledash/553#-JlIe54oB9XXq6v2bsb7)** ([raw file](https://github.com/hammerlab/cycledash/blob/2d190e8e0a9dac68f0a3209ece6e879ac1b023fb/cycledash/static/js/examine/components/CommentBox.js#L209)):</sup>
Done.

---

<sup>**[cycledash/static/js/examine/components/CommentBox.js, line 373 \[r8\]](https://reviewable.io:443/reviews/hammerlab/cycledash/553#-JlLUPyp22PLbJA12tj2)** ([raw file](https://github.com/hammerlab/cycledash/blob/2d190e8e0a9dac68f0a3209ece6e879ac1b023fb/cycledash/static/js/examine/components/CommentBox.js#L373)):</sup>
Why no need? I don't want the `setCommentTextState` below to get called?

---

<sup>**[cycledash/static/js/examine/RecordStore.js, line 214 \[r7\]](https://reviewable.io:443/reviews/hammerlab/cycledash/553#-JlDDphVhXAUwwdDUpW1)** ([raw file](https://github.com/hammerlab/cycledash/blob/f6e0a142f7058a904084d2028af03c9ac68d5797/cycledash/static/js/examine/RecordStore.js#L214)):</sup>
Yeah, I agree that a server-side join probably makes sense, though I think that should be a separate PR.

---

<sup>**[cycledash/static/js/examine/RecordStore.js, line 215 \[r8\]](https://reviewable.io:443/reviews/hammerlab/cycledash/553#-JlLYRjXLbvEHzTGr2jt)** ([raw file](https://github.com/hammerlab/cycledash/blob/2d190e8e0a9dac68f0a3209ece6e879ac1b023fb/cycledash/static/js/examine/RecordStore.js#L215)):</sup>
Well, I still need `if (_.isEmpty(record.comments)) {`, but still neat! Done.

---

<sup>**[cycledash/static/js/examine/RecordStore.js, line 229 \[r8\]](https://reviewable.io:443/reviews/hammerlab/cycledash/553#-JlL_gr5yJtZjyg8daM7)** ([raw file](https://github.com/hammerlab/cycledash/blob/2d190e8e0a9dac68f0a3209ece6e879ac1b023fb/cycledash/static/js/examine/RecordStore.js#L229)):</sup>
I like the usage of `findWhere`! I don't think `findWhere` will be happy with an undefined `record.comments`, so I think that should still be initialized at the top. Also, the clone stuff feels less readable to me.

---

<sup>**[cycledash/static/js/examine/RecordStore.js, line 239 \[r8\]](https://reviewable.io:443/reviews/hammerlab/cycledash/553#-JlLWLXEDWgUY8VbobSr)** ([raw file](https://github.com/hammerlab/cycledash/blob/2d190e8e0a9dac68f0a3209ece6e879ac1b023fb/cycledash/static/js/examine/RecordStore.js#L239)):</sup>
Fixed

---

<sup>**[cycledash/static/js/examine/RecordStore.js, line 334 \[r8\]](https://reviewable.io:443/reviews/hammerlab/cycledash/553#-JlLcaorg1BOmgECTfwE)** ([raw file](https://github.com/hammerlab/cycledash/blob/2d190e8e0a9dac68f0a3209ece6e879ac1b023fb/cycledash/static/js/examine/RecordStore.js#L334)):</sup>
Done. I knew about that function, too, doh!

---

<sup>**[cycledash/static/js/examine/utils.js, line 62 \[r8\]](https://reviewable.io:443/reviews/hammerlab/cycledash/553#-JlLd5cYD2a1QpKsQ6KM)** ([raw file](https://github.com/hammerlab/cycledash/blob/2d190e8e0a9dac68f0a3209ece6e879ac1b023fb/cycledash/static/js/examine/utils.js#L62)):</sup>
I wish (my currently settings in) emacs properly handled this.

---

<sup>**[cycledash/static/js/runs/comments.js, line 1 \[r8\]](https://reviewable.io:443/reviews/hammerlab/cycledash/553#-JlLdLuDGrU_NVJjDPmE)** ([raw file](https://github.com/hammerlab/cycledash/blob/2d190e8e0a9dac68f0a3209ece6e879ac1b023fb/cycledash/static/js/runs/comments.js#L1)):</sup>
Done

---

<sup>**[package.json, line 9 \[r8\]](https://reviewable.io:443/reviews/hammerlab/cycledash/553#-JlLe-9Styl6XjoisS2H)** ([raw file](https://github.com/hammerlab/cycledash/blob/2d190e8e0a9dac68f0a3209ece6e879ac1b023fb/package.json#L9)):</sup>
Ah, silly me. Totally missed that.

---

<sup>**[schema.sql, line 47 \[r8\]](https://reviewable.io:443/reviews/hammerlab/cycledash/553#-JlLeBu86a97bz0hZREA)** ([raw file](https://github.com/hammerlab/cycledash/blob/2d190e8e0a9dac68f0a3209ece6e879ac1b023fb/schema.sql#L47)):</sup>
Looks like you didn't read through the PR text on GitHub ;)

---

<sup>**[tests/data/comment-db.json, line 46 \[r8\]](https://reviewable.io:443/reviews/hammerlab/cycledash/553#-JlLflxkHizTGMY__G6m)** ([raw file](https://github.com/hammerlab/cycledash/blob/2d190e8e0a9dac68f0a3209ece6e879ac1b023fb/tests/data/comment-db.json#L46)):</sup>
Done

---


<!-- Sent from Reviewable.io -->

ihodes commented:
Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/hammerlab/cycledash/553)



---

Good catch in /comments! 

Re: screenshots: I'd just say for you to do what you think best; I don't think this is a critical decision we need to block on. They're a tool to help us.

---

<sup>**[cycledash/static/js/examine/components/CommentBox.js, line 90 \[r8\]](https://reviewable.io:443/reviews/hammerlab/cycledash/553#-JlIa6LAOB4Y9OWzHvay)** ([raw file](https://github.com/hammerlab/cycledash/blob/2d190e8e0a9dac68f0a3209ece6e879ac1b023fb/cycledash/static/js/examine/components/CommentBox.js#L90)):</sup>
Ah that was silly of me, it could also stay on the object like so: `LOCAL_STORAGE_AUTHOR_KEY: 'CYCLEDASH_AUTHORNAME'`, but obviously this is fine too :)

---

<sup>**[cycledash/static/js/examine/components/CommentBox.js, line 94 \[r8\]](https://reviewable.io:443/reviews/hammerlab/cycledash/553#-JlI_vTO52KrKHBjQsd3)** ([raw file](https://github.com/hammerlab/cycledash/blob/2d190e8e0a9dac68f0a3209ece6e879ac1b023fb/cycledash/static/js/examine/components/CommentBox.js#L94)):</sup>
Yikes, good to know!

---

<sup>**[cycledash/static/js/examine/components/CommentBox.js, line 129 \[r8\]](https://reviewable.io:443/reviews/hammerlab/cycledash/553#-JlIbfC8y3hCOPkkohsG)** ([raw file](https://github.com/hammerlab/cycledash/blob/2d190e8e0a9dac68f0a3209ece6e879ac1b023fb/cycledash/static/js/examine/components/CommentBox.js#L129)):</sup>
I'm fine with either, for sure. I like the rule of thumb that var name length should be somewhat proportional to the scope it's being used in—I think @smondet articulated that in some OCaml presentation. (But comment is totally 100% fine as well, don't change it)

---

<sup>**[cycledash/static/js/examine/components/CommentBox.js, line 183 \[r8\]](https://reviewable.io:443/reviews/hammerlab/cycledash/553#-JlIdUzvXmyQcW6eUfUR)** ([raw file](https://github.com/hammerlab/cycledash/blob/2d190e8e0a9dac68f0a3209ece6e879ac1b023fb/cycledash/static/js/examine/components/CommentBox.js#L183)):</sup>
Ah yeah good, point, sorry! Carry on. `defaultEditState` makes more sense knowing that. `startInEditState` seems good, too.

---

<sup>**[cycledash/static/js/examine/components/CommentBox.js, line 373 \[r8\]](https://reviewable.io:443/reviews/hammerlab/cycledash/553#-JlLUPyp22PLbJA12tj2)** ([raw file](https://github.com/hammerlab/cycledash/blob/2d190e8e0a9dac68f0a3209ece6e879ac1b023fb/cycledash/static/js/examine/components/CommentBox.js#L373)):</sup>
Oh ugh Reviewable had hidden the below code from me, so I assumed that was a closing brace for the function—sorry about that.

---

<sup>**[cycledash/static/js/examine/RecordStore.js, line 214 \[r7\]](https://reviewable.io:443/reviews/hammerlab/cycledash/553#-JlDDphVhXAUwwdDUpW1)** ([raw file](https://github.com/hammerlab/cycledash/blob/f6e0a142f7058a904084d2028af03c9ac68d5797/cycledash/static/js/examine/RecordStore.js#L214)):</sup>
Do you get the sense it's a bigger change? If so, I agree it makes sense to sequester it in a different PR if it'll take a while. File it?

---

<sup>**[cycledash/static/js/examine/RecordStore.js, line 215 \[r8\]](https://reviewable.io:443/reviews/hammerlab/cycledash/553#-JlLYRjXLbvEHzTGr2jt)** ([raw file](https://github.com/hammerlab/cycledash/blob/2d190e8e0a9dac68f0a3209ece6e879ac1b023fb/cycledash/static/js/examine/RecordStore.js#L215)):</sup>
Is there a reason you need to delete `record.comments` if it's empty? Seems like you don't need it. 

---

<sup>**[cycledash/static/js/examine/RecordStore.js, line 224 \[r8\]](https://reviewable.io:443/reviews/hammerlab/cycledash/553#-JlL_gr5yJtZjyg8daM7)** ([raw file](https://github.com/hammerlab/cycledash/blob/2d190e8e0a9dac68f0a3209ece6e879ac1b023fb/cycledash/static/js/examine/RecordStore.js#L224)):</sup>
The findWhere will work (return null) with undefined or null or empty record.comments, so no need to initialize it if it's empty. The joys of functional programming! But it's fine too, and you'd be doing it before appending to record.comments below anyway. Fair enough re clone, I like your way here :)

---

<sup>**[cycledash/static/js/examine/utils.js, line 62 \[r8\]](https://reviewable.io:443/reviews/hammerlab/cycledash/553#-JlLd5cYD2a1QpKsQ6KM)** ([raw file](https://github.com/hammerlab/cycledash/blob/2d190e8e0a9dac68f0a3209ece6e879ac1b023fb/cycledash/static/js/examine/utils.js#L62)):</sup>
I know that feeling; I think my mode makes me do this manually, too.

---

<sup>**[schema.sql, line 47 \[r8\]](https://reviewable.io:443/reviews/hammerlab/cycledash/553#-JlLeBu86a97bz0hZREA)** ([raw file](https://github.com/hammerlab/cycledash/blob/2d190e8e0a9dac68f0a3209ece6e879ac1b023fb/schema.sql#L47)):</sup>
:( I did, promise! 

(Sorry)

---

<sup>**[tests/data/comment-db.json, line 46 \[r8\]](https://reviewable.io:443/reviews/hammerlab/cycledash/553#-JlLflxkHizTGMY__G6m)** ([raw file](https://github.com/hammerlab/cycledash/blob/2d190e8e0a9dac68f0a3209ece6e879ac1b023fb/tests/data/comment-db.json#L46)):</sup>
Ah was this just in the comment-db.json, not in other code?

---


<!-- Sent from Reviewable.io -->

tavinathanson committed:
Fix react key issue + tests

tavinathanson committed:
Fix records vs. record bug

tavinathanson committed:
Address more code review comments

tavinathanson commented:
Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/hammerlab/cycledash/553)







---

<sup>**[cycledash/static/js/examine/components/CommentBox.js, line 93 \[r8\]](https://reviewable.io:443/reviews/hammerlab/cycledash/553#-JlIa6LAOB4Y9OWzHvay)** ([raw file](https://github.com/hammerlab/cycledash/blob/2d190e8e0a9dac68f0a3209ece6e879ac1b023fb/cycledash/static/js/examine/components/CommentBox.js#L93)):</sup>
Haha! I really wanted to make it more local, but couldn't think of a simple way. I momentarily forgot that the whole thing is just an object. Sigh. Done!

---

<sup>**[cycledash/static/js/examine/RecordStore.js, line 214 \[r7\]](https://reviewable.io:443/reviews/hammerlab/cycledash/553#-JlDDphVhXAUwwdDUpW1)** ([raw file](https://github.com/hammerlab/cycledash/blob/f6e0a142f7058a904084d2028af03c9ac68d5797/cycledash/static/js/examine/RecordStore.js#L214)):</sup>
Well, big enough that I think it should be it's own PR :)

https://github.com/hammerlab/cycledash/issues/561

For what it's worth, this particular function would still be here because of (as Dan said) optimistic updating. 

---

<sup>**[cycledash/static/js/examine/RecordStore.js, line 215 \[r8\]](https://reviewable.io:443/reviews/hammerlab/cycledash/553#-JlLYRjXLbvEHzTGr2jt)** ([raw file](https://github.com/hammerlab/cycledash/blob/2d190e8e0a9dac68f0a3209ece6e879ac1b023fb/cycledash/static/js/examine/RecordStore.js#L215)):</sup>
Whether or not a comment bubble was shown was relying on that, but I changed that too. Done.

---

<sup>**[cycledash/static/js/examine/RecordStore.js, line 224 \[r8\]](https://reviewable.io:443/reviews/hammerlab/cycledash/553#-JlL_gr5yJtZjyg8daM7)** ([raw file](https://github.com/hammerlab/cycledash/blob/2d190e8e0a9dac68f0a3209ece6e879ac1b023fb/cycledash/static/js/examine/RecordStore.js#L224)):</sup>
Good to know, though I'll leave it as is!

---

<sup>**[tests/data/comment-db.json, line 46 \[r8\]](https://reviewable.io:443/reviews/hammerlab/cycledash/553#-JlLflxkHizTGMY__G6m)** ([raw file](https://github.com/hammerlab/cycledash/blob/2d190e8e0a9dac68f0a3209ece6e879ac1b023fb/tests/data/comment-db.json#L46)):</sup>
No, I also checked the corresponding Python code :)

Look for `value.isoformat()`

---

<sup>**[tests/pdifftests/test.py, line 46 \[r7\]](https://reviewable.io:443/reviews/hammerlab/cycledash/553#-JlDBc5q_Mau7JHK31VG)** ([raw file](https://github.com/hammerlab/cycledash/blob/f6e0a142f7058a904084d2028af03c9ac68d5797/tests/pdifftests/test.py#L46)):</sup>
I kept the dimensions the same and used phantom1 for now. Note that the dimensions still change because there are more columns (thanks to one of Isaac's changes).

---


<!-- Sent from Reviewable.io -->

tavinathanson commented:
@ihodes back to you, let me know when it's mergeable!

ihodes commented:
Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/hammerlab/cycledash/553)







---

<sup>**[cycledash/static/js/examine/RecordStore.js, line 214 \[r7\]](https://reviewable.io:443/reviews/hammerlab/cycledash/553#-JlDDphVhXAUwwdDUpW1)** ([raw file](https://github.com/hammerlab/cycledash/blob/f6e0a142f7058a904084d2028af03c9ac68d5797/cycledash/static/js/examine/RecordStore.js#L214)):</sup>
Makes sense to me :) 

---

<sup>**[cycledash/static/js/examine/RecordStore.js, line 215 \[r8\]](https://reviewable.io:443/reviews/hammerlab/cycledash/553#-JlLYRjXLbvEHzTGr2jt)** ([raw file](https://github.com/hammerlab/cycledash/blob/2d190e8e0a9dac68f0a3209ece6e879ac1b023fb/cycledash/static/js/examine/RecordStore.js#L215)):</sup>
Ah sorry, I didn't realize that. 

---


<!-- Sent from Reviewable.io -->

ihodes commented:
@tavinathanson merge away!
